R/layout.R

```diff
@@ -290,291 +287,93 @@ layout_columns <- function(
   )
 }
 
+as_col_spec <- function(col_widths, n_kids) {
+  if (is.null(col_widths)) return(NULL)
 
-
-impute_col_spec <- function(x, n_kids) {
-  if (isTRUE(is.na(x))) {
-    x <- breakpoints(sm = NA, lg = NA)
+  if (!is_breakpoints(col_widths)) {
+    col_widths <- breakpoints(md = col_widths)
   }
 
-  if (!is_breakpoints(x)) {
-    x <- breakpoints(md = x)
-  }
+  for (break_name in names(col_widths)) {
+    bk <- col_widths[[break_name]]
 
-  # if smallest defined breakpoint is large, fill in gap with 'md'
-  if (names(x)[1] %in% c("lg", "xl", "xxl")) {
-    x <- breakpoints(md = x[[1]], !!!x)
-  }
+    if (rlang::is_na(bk)) {
+      next
+    }
 
-  has_auto_spec <- vapply(x, function(y) isTRUE(is.na(y)), logical(1))
+    if (isTRUE(any(bk == 0))) {
+      abort("Column values must be greater than 0 to indicate width, or negative to indicate a column offset.")
+    }
 
-  if (!any(has_auto_spec)) {
-    return(list(n_cols = 12, col_widths = x))
-  }
+    if (length(bk) > 1 && anyNA(bk)) {
+      abort("Cannot mix widths and `NA` values. All column widths must be specified, or choose auto widths using a single `NA` value.")
+    }
 
-  # If the user gave us column widths, we have to use 12 columns at all
-  # breakpoints. If they didn't, we let the best fit algorithm pick.
-  # `n_col_basis` is stable for all breakpoints, `n_cols` is the final number.
-  n_cols <- n_col_basis <- if (any(!has_auto_spec)) 12 else NA
-
-  for (break_name in names(x)[has_auto_spec]) {
-    best_fit <- col_width_best_fit(
-      n_kids,
-      prefer_wider = break_name %in% c("sm", "md"),
-      n_cols = n_col_basis
-    )
-
-    x[[break_name]] <- best_fit$col_widths
-
-    if (is.na(n_cols)) {
-      # The best fit algorithm picks the same `n_cols` given `n_items`,
-      # so we use the first one as the final `n_cols` for all breakpoints. In
-      # the future `n_cols` could be breakpoint-specific as well.
-      n_cols <- best_fit$n_cols
+    if (!any(bk > 0)) {
+      abort("Column values must include at least one positive integer width.")
+    }
+
+    if (length(bk) > n_kids) {
+      rlang::warn(
+        sprintf(
+          "More column widths than children at breakpoint '%s', extra widths will be ignored.",
+          break_name
+        )
+      )
     }
   }
 
-  list(n_cols = n_cols, col_widths = x)
+  col_widths
 }
 
+json_col_spec <- function(col_widths) {
+  if (is.null(col_widths)) return(NULL)
 
+  jsonlite::toJSON(
+    unclass(col_widths),
+    null = "null",
+    na = "null",
+    auto_unbox = FALSE
+  )
+}
 
 row_heights_css_vars <- function(x) {
   if (is.null(x)) return(list())
 
+  maybe_fr_unit <- function(x) {
+    if (is.numeric(x)) paste0(x, "fr") else x
+  }
+
   if (!is_breakpoints(x)) {
-    x <- breakpoints(sm = x)
+    x <- breakpoints(xs = x)
   }
```

Comment:
The thinking here is that `"xs"` is equivalent to "for all breakpoints" and we'll still want to handle `breakpoints(xs = c(1, 2), ...)`, since the user may want to set the row heights from `xs` to a higher breakpoint, for example.

If that happens, we ultimately do the same thing that we would do for a non-breakpoint `row_heights`, so we might as well combine the two cases rather than repeating the same code twice.

Reply:
Let's consider adding `xs` to `breakpoints()` and supporting it for `col_widths`